osu.Game.Rulesets.Osu/Mods/OsuModDepth.cs

```diff
@@ -0,0 +1,132 @@
+
+        protected override void ApplyNormalVisibilityState(DrawableHitObject hitObject, ArmedState state) => applyTransform(hitObject, state);
+
+        public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset)
+        {
+            // Hide judgment displays and follow points as they won't make any sense.
+            // Judgements can potentially be turned on in a future where they display at a position relative to their drawable counterpart.
+            drawableRuleset.Playfield.DisplayJudgements.Value = false;
+            (drawableRuleset.Playfield as OsuPlayfield)?.FollowPoints.Hide();
+        }
+
+        private void applyTransform(DrawableHitObject drawable, ArmedState state)
+        {
+            switch (drawable)
+            {
+                case DrawableHitCircle circle:
+                    if (!ShowApproachCircles.Value)
+                    {
+                        var hitObject = (OsuHitObject)drawable.HitObject;
+                        double appearTime = hitObject.StartTime - hitObject.TimePreempt;
+
+                        using (circle.BeginAbsoluteSequence(appearTime))
+                            circle.ApproachCircle.Hide();
+                    }
+
+                    break;
+            }
+        }
+
+        public void Update(Playfield playfield)
+        {
+            double time = playfield.Time.Current;
+
+            foreach (var drawable in playfield.HitObjectContainer.AliveObjects)
+            {
+                switch (drawable)
+                {
+                    case DrawableHitCircle circle:
+                        processObject(time, circle, 0);
+                        break;
+
+                    case DrawableSlider slider:
+                        processObject(time, slider, slider.HitObject.Duration);
+                        break;
+                }
+            }
+        }
+
+        private void processObject(double time, DrawableOsuHitObject drawable, double duration)
+        {
+            var hitObject = drawable.HitObject;
+
+            double baseSpeed = MaxDepth.Value / hitObject.TimePreempt;
+            double offsetAfterStartTime = duration + hitObject.MaximumJudgementOffset + 500;
+            double slowSpeed = Math.Min(-minDepth / offsetAfterStartTime, baseSpeed);
+
+            double decelerationTime = hitObject.TimePreempt * 0.2;
+            float decelerationDistance = (float)(decelerationTime * (baseSpeed + slowSpeed) * 0.5);
+
+            float z;
+
+            if (time < hitObject.StartTime)
+            {
+                double timeOffset = time - (hitObject.StartTime - decelerationTime);
+                double deceleration = (slowSpeed - baseSpeed) / decelerationTime;
+                z = decelerationDistance - (float)(baseSpeed * timeOffset + deceleration * timeOffset * timeOffset * 0.5);
+            }
+            else
```

Comment:
My proposal was to only slow objects down when actually required (even if applied to all sliders to start with). I think not applying this whenever we can avoid doing so will make the visuals of this mod much better.

Reply:
Ah okay, we are on the same page then

Reply:
Even better would be only applying when a slider gets "too long" (maybe too long for autoplay to hit it correctly with no autoplay path adjustment?)

Reply:
So I've made some changes. Now every object moves at a constant speed, except for long sliders. There's a check for how big it would become at the end time if allowed to move at a constant speed. If its scale will be bigger than 1.5 deceleration will be applied.